dds/DCPS/XTypes/DynamicType.cpp

```diff
@@ -0,0 +1,138 @@
+{
+  descriptor = descriptor_;
+}
+
+TypeDescriptor_rch DynamicType::get_descriptor()
+{
+  return descriptor_;
+}
+
+DCPS::String DynamicType::get_name() const
+{
+  return descriptor_->name;
+}
+
+TypeKind DynamicType::get_kind() const
+{
+  return descriptor_->kind;
+}
+
+DDS::ReturnCode_t DynamicType::get_member_by_name(DynamicTypeMember_rch& member, const DCPS::String& name) const
+{
+  const DynamicTypeMembersByName::const_iterator pos = member_by_name.find(name);
+  if (pos != member_by_name.end()) {
+    member = pos->second;
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+void DynamicType::get_all_members_by_name(DynamicTypeMembersByName& member)
+{
+  member = member_by_name;
+}
+
+DDS::ReturnCode_t DynamicType::get_member(DynamicTypeMember_rch& member, MemberId id) const
+{
+  DynamicTypeMembersById::const_iterator pos = member_by_id.find(id);
+  if (pos != member_by_id.end()) {
+    member = pos->second;
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+void DynamicType::get_all_members(DynamicTypeMembersById& member)
+{
+  member = member_by_id;
+}
+
+ACE_CDR::ULong DynamicType::get_member_count() const
+{
+  return CORBA::ULong(member_by_index.size());
+}
+
+DDS::ReturnCode_t DynamicType::get_member_by_index(DynamicTypeMember_rch& member, ACE_CDR::ULong index) const
+{
+  if (index < member_by_index.size()) {
+    member = member_by_index[index];
+    return DDS::RETCODE_OK;
+  } else {
+    return DDS::RETCODE_ERROR;
+  }
+}
+
+bool DynamicType::equals(const DynamicType& other) const
+{
+  //7.5.2.8.4 Operation: equals
+  //Two types shall be considered equal if and only if all of their respective properties, as identified
+  //in Table 54 above, are equal.
+
+  //In addition to what the spec says to compare, we will be comparing the TypeDescriptors of both
+  //DynamicTypes. The spec seems to assume this is the case, despite not listing the TypeDescriptor
+  //as a property in table 54. This is done to allow the recursive comparison required by the changes
+  //to DynamicTypeMember::equals.
+  return *this == other;
+}
+
+void DynamicType::insert_dynamic_member(const DynamicTypeMember_rch& dtm)
+{
+  member_by_index.push_back(dtm);
+  if (dtm->get_descriptor()->id != MEMBER_ID_INVALID) {
+    member_by_id.insert(std::make_pair(dtm->get_descriptor()->id , dtm));
+  }
+  member_by_name.insert(std::make_pair(dtm->get_descriptor()->name , dtm));
+}
+
+bool DynamicType::test_equality_i(const DynamicType& rhs, DynamicTypePtrPairSeen& dt_ptr_pair) const
+{
+  //check pair seen
+  DynamicTypePtrPair this_pair = std::make_pair(this, &rhs);
+  DynamicTypePtrPairSeen::const_iterator have_seen = dt_ptr_pair.find(this_pair);
+  if (have_seen == dt_ptr_pair.end()) {
+    dt_ptr_pair.insert(this_pair);
+    return
+      OpenDDS::XTypes::test_equality_i(*this->descriptor_.in(), *rhs.descriptor_.in(), dt_ptr_pair) &&
```

Comment:
```suggestion
      test_equality_i(*descriptor_, *rhs.descriptor_, dt_ptr_pair) &&
```

Reply:
test_equality_i exists both in the class namespace as well as in OpenDDS::XTypes.  These calls specifically need the versions from the OpenDDS::XTypes namespace, which they won't find without the scoping.